### PR TITLE
[Relax] Add CopyWithNewParams interface

### PR DESCRIPTION
**Pull request by Ubospica**
This is the new PR following #90 after the source branch having been moved to personal repositiory.

This PR adds a new interface `relax.utils.copy_with_new_params` to copy functions:
```
def copy_with_new_params(func: Function) -> Function:
    pass
```

The parameters of the original function would be copied to satisfy the restriction in the well-formed check: any two functions should not share the same parameter variable.

This interface serves for passes that needs to copy some existing function and modify it, such as `relax.transform.Gradient`.

### CODE REVIEW

**Comment by SiriusNEO** [comment]
I'm thinking if the struct info of params contains Var, should we copy them as well?
For example, the shape of an input Tensor is a relax.Var.

**Comment by Ubospica** [comment]
> I'm thinking if the struct info of params contains Var, should we copy them as well? For example, the shape of an input Tensor is a relax.Var.

@SiriusNEO I perfer not. Because we now only have to satisfy the requirement in well-formed check, which only checks parameter vars of functions. If we wanna copy all vars in the function, we may have to do this.

File: src/relax/utils.cc [hunk 1]
```diff
@@ -84,5 +84,29 @@ bool IsLeafExpr(const Expr& expr) {
...
+    for (Var param : func->params) {
+      Var new_param = Var(param->vid, GetStructInfo(param), param->span);
+      this->var_remap_[param->vid] = new_param;
+      new_params.push_back(new_param);
+    }
+
+    Expr body = this->VisitWithNewScope(func->body, new_params);
+
+    return Function(new_params, body, func->ret_struct_info, func->attrs);
+  }
```
Comment (MasterJH5574) [inline src/relax/utils.cc]: Consider the case where the Function you want to copy contains another local Function inside. If we override the visit method, we will recurse into the local function, which is not as expected. So let’s change the visit method to a public method, which takes a Function, creates a new set of params, sets the var remapping and recurses into the body.

**Comment by MasterJH5574** [comment]
> I'm thinking if the struct info of params contains Var, should we copy them as well?
> For example, the shape of an input Tensor is a relax.Var.

The behavior and semantics of param struct info containing shape var needs more thinking, and currently we have no such example. We will revisit this point after some time and at this moment let’s just go as usual, assuming no shape var appears in the params’ structure info.

File: tests/python/relax/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import pytest
+import tvm
+import tvm.script
+from tvm import relax
+from tvm.ir.base import assert_structural_equal
+from tvm.script.parser import ir as I, relax as R, tir as T
```
Comment (MasterJH5574) [inline tests/python/relax/test_utils.py]: Minor nit: seems that `I` and `T` are not used.

**Review by MasterJH5574 — APPROVED** [review]

**Pull request merged**